include/cocaine/rpc/asio/header.hpp

```diff
@@ -0,0 +1,558 @@
+/*
+    Copyright (c) 2011-2015 Anton Matveenko <antmat@yandex-team.ru>
+    Copyright (c) 2011-2015 Other contributors as noted in the AUTHORS file.
+
+    This file is part of Cocaine.
+
+    Cocaine is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    Cocaine is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef HEADER_HPP
+#define HEADER_HPP
+
+#include "cocaine/traits.hpp"
+#include "cocaine/exceptions.hpp"
+
+#include <cstring>
+#include <array>
+#include <cassert>
+#include <functional>
+
+namespace cocaine { namespace io  {
+
+enum class header_id {
+    TRACE_ID,
+    SPAN_ID,
+    PARENT_ID,
+    //Not used. Just indicates size of headers.
+    LAST
+};
+
+// Represents header name. Can be constructed only from headers enum.
+class header_key_t {
+public:
+    constexpr
+    header_key_t(header_id _id) :
+        id(_id)
+    {}
+
+    inline
+    const std::string&
+    name() const {
+        return names()[static_cast<size_t>(id)];
+    }
+
+private:
+    header_id id;
+    static constexpr size_t header_cnt = static_cast<size_t>(header_id::LAST);
+
+    inline
+    static
+    const std::array<std::string, header_cnt>&
+    names() {
+        static std::array<std::string, header_cnt> header_names {{
+            "trace_id",
+            "span_id",
+            "parent_id"
+        }};
+        return header_names;
+    }
+};
+
+// Header class.
+// Represents non-owning header referring to some external memory.
+// Can be constructed only via header table
+class header_t {
+public:
+    struct str {
+        const char* data;
+        size_t size;
+        bool operator==(const str& other) const {
+            return size == other.size && (memcmp(data, other.data, size) == 0);
+        }
+    };
+
+    header_t(const header_t& other) = default;
```

Comment:
This one will result in warning on some ancient compilers when tried to emplace copy ctor somewhere.